### DISCUSSION TITLE
Query to Return Rows with Id's passed from client.

### DISCUSSION THREAD

**Discussion by stixoffire** [serenity-is/Serenity#6448] [Help] [answered]
I have a need to return entities where the ID's are in a list of such as  Int[] . And then I need to add them to the InMemory Grid.

Currently I seem to be stuck on calling the retrieve handler and passing the Id's which seems like I need to create a custom method in the repository . 

        public ListResponse<MyRow> List(IDbConnection connection, ListRequestByCriteria request)
        { 
        
        }
        
        ListRequestByCriteria  has simply derive from request and has KeyIds[] as a property.
        
        public class ListRequestByCriteria : Serenity.Services.ListRequest
        {        
            public Int32[] Keys { get; set; } 
         }

So how do I use this in my repository method to return a list of entities where the ID's are in the keyIds[] ?

Then I want to use this to add to an in memory grid.

**Comment by petebrieso** [comment]
for first part just have your repository call the regular List but pass the id's in the Criteria part of the list request
`new ListRequest { Criteria = new Criteria("Id").In(Keys) }`

**Reply by stixoffire** [reply]
I am calling this from the client side. I get cannot find name ListRequest, no value exists in scope for shorthand property criteria.





MyService.List(new ListRequest { Criteria = new Criteria("TrainingId").In(Keys) }, response =>{});


From what I have read you can not apply SQL query from client side but must use applyfilters in the Handler .

Do you have any working code example calling a service from the client with parameters  and returning  entities ?

**Reply by petebrieso** [reply]
If you're calling from client you can just call the basic list service and pass "in" criteria like this:
```
         xyzService.List({
                Criteria: [[xyzRow.Fields.Id], "in", '(1,2,3)'] //1,2,3 represents your list of id's but note they have to be in braces.
            }, response => {
                response.Entities.forEach(row => {
                    console.log(row.Name);  //do something with each row here
                })
            });
```

**Reply by stixoffire** [reply]
Ah some caveats Criteria:  with Criteria in brackets and values in Parenthesis . 

I will try that .. Thank you..

**Reply by stixoffire** [reply]
I tried it and I get incorrect syntax near @P1 - when I check the debug script it @p1 is showing exactly like this '( 2, 4)' 
If I put in SSMS and set the query and remove the apostrophe's **'**  so I have this ( 2, 4) that will allow the query to function.

Code is below .. so how do I stop it from throwing in the apostrophe's ?
 
Later I will try this 
InKeys = "IN (" + InKeys.substring(1, InKeys.length) + ")";
and something like ..
[[TrainingRow.Fields.TrainingId], InKeys , null]

```
protected myMethod(keyIds: string[]) {

            keyIds.forEach(key => { InKeys = InKeys + ", "  + key  });
            InKeys = "(" + InKeys.substring(1, InKeys.length) + ")";

XYZService.List({
                Criteria: [[XYZRow.Fields.TrainingId], "In", InKeys] 
}, response => {
                response.Entities.forEach(row => {
                    console.log(row.Name);  
                })
            });
}
```

**Reply by petebrieso** [reply]
you can more easily make a csv from array with `keyIds.join(",")`

**Reply by petebrieso** [reply]
make sure you use "in" as the second parameter, note the lower-case

**Reply by stixoffire** [reply]
thanks for the csv tip ..
"in" is the second parameter that seems fine - when this translates to the SQL - in the debug output the script is always putting InKeys with 
single quotes **'** (1,4)**'** 

That is the issue I am having it translates it and wants to quote the variable.

I have tried but none of these are working . it sets @p1
```
Criteria: [[XYZRow.Fields.TrainingId], "in", InKeys] 
Criteria: [[XYZRow.Fields.TrainingId], "in", InKeys.valueOf()] 
Criteria: [[XYZRow.Fields.TrainingId], "in", [InKeys]] 
Criteria: [[XYZRow.Fields.TrainingId], "in", (InKeys)] 
Criteria: [[XYZRow.Fields.TrainingId], "in", {InKeys}] 
```

The query keeps resolving to 
```
DECLARE @p1 NVARCHAR(4000) = '(1, 2)';
DECLARE @p2 INT = 0;
SELECT T0.[TrainingId] AS [TrainingId], T0.[Name] AS [Name]  
WHERE ((T0.[TrainingId] IN @p1))
```

The problem is how it identifies @P1 - the Single Quotes..  - if I run that query manually in my Query Analyzer it works with out those single quotes but not with them.

**Reply by stixoffire** [reply]
This is the exception 
System.Data.SqlClient.SqlException: 'Incorrect syntax near '@p1'.'

**Reply by stixoffire** [reply]
Notice the single quotes ..

**Reply by VictorTomaili** [reply]
`[[XYZRow.Fields.TrainingId], Serenity.Criteria.Operator.in, [['10','25']]]`
this means 
`[[XYZRow.Fields.TrainingId], Serenity.Criteria.Operator.in, [InKeys.map(x => x.toString())]]`

Doubled arrays are for detecting the real array in criteria because of the criteria itself is already an array.

**Comment by stixoffire** [comment, marked as answer]
Thank you @VictorTomaili !

[VictorTomaili](https://github.com/VictorTomaili)
 had the answer - so for me to  mark this as the answer I put it here.
[[XYZRow.Fields.TrainingId], Serenity.Criteria.Operator.in, [['10','25']]]
this means
[[XYZRow.Fields.TrainingId], Serenity.Criteria.Operator.in, [InKeys.map(x => x.toString())]]

Doubled arrays are for detecting the real array in criteria because of the criteria itself is already an array.
